### ISSUE TITLE
Directional Vectors

### ISSUE THREAD

**Issue by sheshap** [guochengqian/PointNeXt#90]
Dear Authors,

Can you please verify and confirm if there is a bug?

Tensor after grouping: torch.Size([32, 3, 512, 32]) after [Line](https://github.com/guochengqian/openpoints/blob/ee100c81b1d9603c0fc76a3ee4e37d10b2af60ba/models/layers/group.py#L249)
After ball querying around an anchor point, the sampled neighborhood points are:

[[-0.43345895  1.0695219   0.23178683]
 [-0.4364147   1.1959834   0.3461398 ]
 [-0.32737926  1.2161976   0.20860705]
 [-0.37404662  1.2161976   0.15377748]
 [-0.44473666  1.1490581   0.12246026]
 [-0.45984167  1.2161976   0.13207951]
 [-0.43345895  1.1510899   0.29191628]
 [-0.43345895  1.1312102   0.2010018 ]
 [-0.46045053  1.2161976   0.21825439]
 [-0.39841616  1.2161976   0.28159994]
 [-0.41827857  1.2161976   0.18159038]
 [-0.43345895  1.0695219   0.23178683]
 [-0.43345895  1.0695219   0.23178683]
 [-0.43345895  1.0695219   0.23178683]
...
 [-0.43345895  1.0695219   0.23178683]]

However, once the anchor points are subtracted from these neighbor points (to get directional vectors). After [Line](https://github.com/guochengqian/openpoints/blob/ee100c81b1d9603c0fc76a3ee4e37d10b2af60ba/models/layers/group.py#L251) the directional vectors are as follows.
[[ 0.02699158 -0.1466757   0.01353244]
 [ 0.02403584 -0.0202142   0.1278854 ]
 [ 0.13307127  0.         -0.00964734]
 [ 0.08640391  0.         -0.06447691]
 [ 0.01571387 -0.06713951 -0.09579413]
 [ 0.00060886  0.         -0.08617488]
 [ 0.02699158 -0.0651077   0.07366189]
 [ 0.02699158 -0.0849874  -0.01725259]
 [ 0.          0.          0.        ]
 [ 0.06203437  0.          0.06334555]
 [ 0.04217196  0.         -0.03666401]
 [ 0.02699158 -0.1466757   0.01353244]
 [ 0.02699158 -0.1466757   0.01353244]
 [ 0.02699158 -0.1466757   0.01353244]
 ...
 [ 0.02699158 -0.1466757   0.01353244]]

Though the anchor point was [-0.43345895  1.0695219   0.23178683] instead, another point  [-0.46045053  1.2161976   0.21825439] was used to subtract.

P.S: This behavior is not seen for the first anchor point.

Please check.

Thanks and Regards,
Shivanand

**Comment by sheshap** [comment]
Hi Authors,

Can you please check this?

Thanks in Advance

**Comment by sheshap** [comment]
@yuchenlichuck Can you please check this issue?

**Comment by guochengqian** [comment]
could you make your example shorter? It would be easier for me to understand. Could you try just five points. and make the line you are referring to **bold**.

**Comment by sheshap** [comment]
In the description above, I am using only one anchor point and its neighborhood points (nsample=32).
I have shorted my problem statement by half now. Please check.

**Comment by ZhengGongzi** [comment]
@sheshap Yes, I had a similar problem with the Ball_Query module. I printed the **_Neigh_idx_** and found that the index of the anchor point is many times not the first (Neigh_idx[0]), but the second or third (Neigh_idx[1] or Neigh_idx[2])). This leads to the use of incorrect anchor coordinates when finding relative coordinates. Fortunately, however, the indexes of all points except the anchor point are still in ascending order of distance, so even if the incorrect anchor point is used, the deviation of coordinates is not too big.
In order to use the correct anchor coordinates, my solution at that time was to splice the anchor indexes directly before  **_Neigh_idx_** in addition to the anchor indexes. This can ensure that the index of the anchor point must be the first index.

**Comment by sheshap** [comment]
@ZhengGongzi Thanks for your input. Would your approach still give the reported accuracy in the paper?

**Comment by ZhengGongzi** [comment]
@sheshap I only used the ball query module in my experiments. I verified the module and my experiments concluded that the neighborhood index is correct, except that the anchor point index is not always in the first position, which I suspect is related to some underlying mechanism of the .cu file. I think this is easy to handle, because the anchor coordinates are available, just make sure you can subtract the correct anchor coordinates, and the ordering of the points shouldn't have an effect on the results.

**Comment by sheshap** [comment]
@ZhengGongzi I see what you are mentioning. The problem here is are we repeating the right anchor point? and then subtract it from neighborhood points. In the above example it is not the case. I don't mind where the anchor point is but if we do not use right anchor point to subtract then we are having incorrect directional vectors representing a given neighborhood.

**Comment by guochengqian** [comment]
Thanks for the observations and solutions.

**Comment by guochengqian** [comment]
Closing as answered. The directional-vector/anchor-index behavior was discussed in the thread; please open a fresh issue with a minimal current-code repro if there is still an actionable bug.
